tutorials/plot_background_filtering.py

```diff
@@ -0,0 +1,365 @@
+
+As outlined in [1]_, FIR and IIR have different tradeoffs:
+
+    * A causal FIR filter can be linear-phase -- i.e., the same time delay
+      across all frequencies -- whereas a causal IIR filter cannot. The phase
+      and group delay characteristics are also usually better for FIR filters.
+    * IIR filters can generally have a steeper cutoff than an FIR filter of
+      equivalent order.
+    * IIR filters are generally less numerically stable, in part due to
+      accumulating error (due to its recursive calculations).
+
+When designing a filter (FIR or IIR), there are always tradeoffs that
+need to be considered, including but not limited to:
+
+    1. Ripple in the pass-band
+    2. Attenuation of the stop-band
+    3. Steepness of roll-off
+    4. Filter order (i.e., length for FIR filters)
+    5. Time-domain ringing
+
+In general, the sharper something is in frequency, the broader it is in time,
+and vice-versa. This is a fundamental time-frequency tradeoff, and it will
+show up below.
+
+Here we will focus first on FIR filters, which are the default filters used by
+MNE-Python.
+"""
+
+###############################################################################
+# Designing an FIR filter
+# -----------------------
+# First let's import some useful tools for filtering, and set some default
+# values for our data that are reasonable for M/EEG data.
+
+import numpy as np
+from scipy import signal, fftpack
+import matplotlib.pyplot as plt
+
+from mne.time_frequency.tfr import morlet
+
+import mne
+
+sfreq = 1000.
+f_p = 40.
+ylim = [-60, 10]  # for dB plots
+xlim = [2, sfreq / 2.]
+blue = '#1f77b4'
+
+###############################################################################
+# Take for example an ideal low-pass filter, which would give a value of 1 in
+# the pass-band (up to frequency :math:`f_p`) and a value of 0 in the stop-band
+# (down to frequency :math:`f_s`) such that :math:`f_p=f_s=40` Hz here
+# (shown to a lower limit of -60 dB for simplicity):
+
+nyq = sfreq / 2.  # the Nyquist frequency is half our sample rate
+freq = [0, f_p, f_p, nyq]
+gain = [1, 1, 0, 0]
+
+
+def box_off(ax):
+    ax.grid(zorder=0)
+    for key in ('top', 'right'):
+        ax.spines[key].set_visible(False)
+
+
+def plot_ideal(freq, gain, ax):
+    freq = np.maximum(freq, xlim[0])
+    xs, ys = list(), list()
+    for ii in range(len(freq)):
+        xs.append(freq[ii])
+        ys.append(ylim[0])
+        if ii < len(freq) - 1 and gain[ii] != gain[ii + 1]:
+            xs += [freq[ii], freq[ii + 1]]
+            ys += [ylim[1]] * 2
+    gain = 10 * np.log10(np.maximum(gain, 10 ** (ylim[0] / 10.)))
+    ax.fill_between(xs, ylim[0], ys, color='r', alpha=0.1)
+    ax.semilogx(freq, gain, 'r--', alpha=0.25, linewidth=4, zorder=3)
+    xticks = [1, 2, 4, 10, 20, 40, 100, 200, 400]
+    ax.set(xlim=xlim, ylim=ylim, xticks=xticks, xlabel='Frequency (Hz)',
+           ylabel='Amplitude (dB)')
+    ax.set(xticklabels=xticks)
+    box_off(ax)
+
```

Comment:
The one on the CircleCI build is terrible, but even the one from my GitHub comment?